In [99]:
import pandas as pd

In [100]:
df = pd.read_csv("Apple_products_pricing_2020_2026.csv")

In [101]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80000 entries, 0 to 79999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Date               80000 non-null  object 
 1   Platform           80000 non-null  object 
 2   Product_Category   80000 non-null  object 
 3   Model_Name         80000 non-null  object 
 4   Condition          80000 non-null  object 
 5   Launch_Price_USD   80000 non-null  int64  
 6   Launch_Price_INR   80000 non-null  int64  
 7   Current_Price_USD  80000 non-null  float64
 8   Current_Price_INR  80000 non-null  float64
 9   Discount_Pct       80000 non-null  float64
 10  Sale_Event         6649 non-null   object 
 11  Stock_Status       80000 non-null  object 
 12  Rating             80000 non-null  float64
 13  Reviews_Count      80000 non-null  int64  
dtypes: float64(4), int64(3), object(7)
memory usage: 8.5+ MB


In [102]:
df.isnull().sum()

Date                     0
Platform                 0
Product_Category         0
Model_Name               0
Condition                0
Launch_Price_USD         0
Launch_Price_INR         0
Current_Price_USD        0
Current_Price_INR        0
Discount_Pct             0
Sale_Event           73351
Stock_Status             0
Rating                   0
Reviews_Count            0
dtype: int64

In [103]:
df.duplicated().sum()

np.int64(0)

In [104]:
df.describe()

,Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR,Discount_Pct,Rating,Reviews_Count
count,80000.000000,80000.000000,80000.000000,80000.000000,80000.000000,80000.000000,80000.000000
mean,963.959125,94467.994250,782.769855,74628.342439,21.418826,4.450099,2405.995925
std,470.086413,46068.468464,461.673970,45117.872172,16.697310,0.269782,1973.630852
min,329.000000,32242.000000,109.930000,9157.680000,-2.000000,3.800000,10.000000
25%,599.000000,58702.000000,432.930000,41686.835000,6.700000,4.300000,896.000000
50%,799.000000,78302.000000,699.740000,67324.020000,21.300000,4.400000,1893.500000
75%,1199.000000,117502.000000,989.112500,96568.117500,36.800000,4.700000,3383.000000
max,1999.000000,195902.000000,2038.970000,203668.710000,73.100000,4.900000,11526.000000


In [105]:
df.nunique()

Date                  2130
Platform                 2
Product_Category         4
Model_Name              31
Condition                2
Launch_Price_USD         9
Launch_Price_INR         9
Current_Price_USD    49611
Current_Price_INR    79626
Discount_Pct           695
Sale_Event               4
Stock_Status             3
Rating                  12
Reviews_Count         8698
dtype: int64

In [106]:
df.shape

(80000, 14)

In [107]:
print("Rows:",df.shape[0])
print("Columns:",df.shape[1])

Rows: 80000
Columns: 14


In [108]:
df.dtypes

Date                  object
Platform              object
Product_Category      object
Model_Name            object
Condition             object
Launch_Price_USD       int64
Launch_Price_INR       int64
Current_Price_USD    float64
Current_Price_INR    float64
Discount_Pct         float64
Sale_Event            object
Stock_Status          object
Rating               float64
Reviews_Count          int64
dtype: object

In [109]:
df.duplicated().sum()

np.int64(0)

In [110]:
df.columns

Index(['Date', 'Platform', 'Product_Category', 'Model_Name', 'Condition',
       'Launch_Price_USD', 'Launch_Price_INR', 'Current_Price_USD',
       'Current_Price_INR', 'Discount_Pct', 'Sale_Event', 'Stock_Status',
       'Rating', 'Reviews_Count'],
      dtype='object')

In [111]:
df.columns= df.columns.str.strip()

In [112]:
df .columns

Index(['Date', 'Platform', 'Product_Category', 'Model_Name', 'Condition',
       'Launch_Price_USD', 'Launch_Price_INR', 'Current_Price_USD',
       'Current_Price_INR', 'Discount_Pct', 'Sale_Event', 'Stock_Status',
       'Rating', 'Reviews_Count'],
      dtype='object')

In [113]:
df["Product_Category"].duplicated().sum()

np.int64(79996)

In [114]:
df.head()

,Date,Platform,Product_Category,Model_Name,Condition,Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR,Discount_Pct,Sale_Event,Stock_Status,Rating,Reviews_Count
0,2020-09-19,Flipkart,Watch,Apple Watch Series 6 (44mm),New,429,42042,435.81,43322.41,-1.6,NaN,In Stock,4.7,40
1,2020-09-20,Flipkart,Watch,Apple Watch Series 6 (44mm),New,429,42042,436.49,42320.43,-1.7,NaN,Out of Stock,4.6,84
2,2020-09-23,Amazon,Watch,Apple Watch Series 6 (44mm),New,429,42042,422.73,40879.36,1.5,NaN,In Stock,4.4,110
3,2020-09-23,Amazon,Watch,Apple Watch Series 6 (44mm),New,429,42042,425.00,42008.70,0.9,NaN,In Stock,4.8,111
4,2020-09-24,Amazon,Watch,Apple Watch Series 6 (44mm),New,429,42042,436.22,41984.28,-1.7,NaN,In Stock,4.7,35


In [115]:
df.tail()

,Date,Platform,Product_Category,Model_Name,Condition,Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR,Discount_Pct,Sale_Event,Stock_Status,Rating,Reviews_Count
79995,2026-07-31,Flipkart,iPhone,iPhone 17 128GB,New,849,83202,861.39,82894.76,-1.5,NaN,In Stock,4.7,1346
79996,2026-07-31,Flipkart,iPhone,iPhone 17 128GB,New,849,83202,832.50,81159.39,1.9,NaN,In Stock,4.7,802
79997,2026-07-31,Amazon,iPhone,iPhone 17 128GB,Renewed/Refurbished,849,83202,657.30,56336.73,22.6,NaN,In Stock,4.2,1020
79998,2026-07-31,Flipkart,iPhone,iPhone 17 128GB,Renewed/Refurbished,849,83202,709.49,61361.07,16.4,NaN,In Stock,3.9,710
79999,2026-07-31,Flipkart,iPhone,iPhone 17 128GB,New,849,83202,834.95,83035.30,1.7,NaN,Low Stock,4.8,1022


In [116]:
df["Platform"].duplicated().sum()

np.int64(79998)

In [117]:
df["Model_Name"].duplicated().sum()

np.int64(79969)

In [118]:
df.dtypes

Date                  object
Platform              object
Product_Category      object
Model_Name            object
Condition             object
Launch_Price_USD       int64
Launch_Price_INR       int64
Current_Price_USD    float64
Current_Price_INR    float64
Discount_Pct         float64
Sale_Event            object
Stock_Status          object
Rating               float64
Reviews_Count          int64
dtype: object

In [119]:
df["Date"].dtype

dtype('O')

In [120]:
df["Date"]=pd.to_datetime(df["Date"])

In [121]:
df.dtypes

Date                 datetime64[ns]
Platform                     object
Product_Category             object
Model_Name                   object
Condition                    object
Launch_Price_USD              int64
Launch_Price_INR              int64
Current_Price_USD           float64
Current_Price_INR           float64
Discount_Pct                float64
Sale_Event                   object
Stock_Status                 object
Rating                      float64
Reviews_Count                 int64
dtype: object

In [122]:
df["Launch_Price_USD"].describe

<bound method NDFrame.describe of 0        429
1        429
2        429
3        429
4        429
        ... 
79995    849
79996    849
79997    849
79998    849
79999    849
Name: Launch_Price_USD, Length: 80000, dtype: int64>

In [123]:
df[df["Launch_Price_USD"]<0]

,Date,Platform,Product_Category,Model_Name,Condition,Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR,Discount_Pct,Sale_Event,Stock_Status,Rating,Reviews_Count


In [124]:
df[df["Launch_Price_USD"]==0]

,Date,Platform,Product_Category,Model_Name,Condition,Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR,Discount_Pct,Sale_Event,Stock_Status,Rating,Reviews_Count


In [125]:
df[df["Launch_Price_INR"]<0]

,Date,Platform,Product_Category,Model_Name,Condition,Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR,Discount_Pct,Sale_Event,Stock_Status,Rating,Reviews_Count


In [126]:
df[df["Launch_Price_INR"]==0]

,Date,Platform,Product_Category,Model_Name,Condition,Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR,Discount_Pct,Sale_Event,Stock_Status,Rating,Reviews_Count


In [127]:
df[df["Current_Price_USD"]<0]

,Date,Platform,Product_Category,Model_Name,Condition,Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR,Discount_Pct,Sale_Event,Stock_Status,Rating,Reviews_Count


In [128]:
df[df["Current_Price_INR"]<0]

,Date,Platform,Product_Category,Model_Name,Condition,Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR,Discount_Pct,Sale_Event,Stock_Status,Rating,Reviews_Count


In [129]:
df["Launch_Price_USD"].sort_values(ascending=False).head()

62129    1999
7779     1999
32530    1999
32529    1999
68792    1999
Name: Launch_Price_USD, dtype: int64

In [130]:
df["Launch_Price_INR"].sort_values(ascending=False).head()

62129    195902
7779     195902
32530    195902
32529    195902
68792    195902
Name: Launch_Price_INR, dtype: int64

In [131]:
df["Current_Price_USD"].sort_values(ascending=False).head()

21797    2038.97
21296    2038.97
54844    2038.90
55303    2038.90
26040    2038.83
Name: Current_Price_USD, dtype: float64

In [132]:
df["Current_Price_INR"].sort_values(ascending=False).head()

21956    203668.71
45690    203627.08
22036    203574.91
30182    203567.74
47024    203426.03
Name: Current_Price_INR, dtype: float64

In [133]:
df["Product_Category"].value_counts()

Product_Category
iPhone    28589
Mac       18020
Watch     17865
iPad      15526
Name: count, dtype: int64

In [134]:
df["Stock_Status"].value_counts()

Stock_Status
In Stock        55034
Out of Stock    13475
Low Stock       11491
Name: count, dtype: int64

## Findings(Data validation Report)
- Validated the dataset to identify the data quality issues, including data types, invalid values,and inconsistent categories.
  

### Findings
1. Data Type validation
     - checked all column data types using df.info()
     - The Date column successfully converted from object to datetime.
     - Price columns such as (Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR) are stored as a numeric datatypes.
2. Invalid value validation
     - No negative or zero values found in the columns(Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR).
     - The maximum value in Launch_Price_USD is 1999 USD , which appears to be a valid premium Apple product price.
3. Category validation       
     - Checked Stock_status for inconsistent values, but outout has only consistent values there are no inconsistent values.
     - Checked Product_category for inconsistent values for category such as ipad, watch , mac, iphone ,there are no inconsistent values.     

#### Conclusion
The dataset passed the Day 4 validtaion checks. No major data quality issued were identified in the price columns.The dataset is ready for the cleaning plan and standardization steps in day 5.

## Day - 4
- Author cleaning plan (missing values, duplicates, type fixes, category harmonisation)
- Decide keep / drop / impute rule for each affected column
- Freeze cleaning plan v1 and commit to /docs

In [135]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80000 entries, 0 to 79999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Date               80000 non-null  datetime64[ns]
 1   Platform           80000 non-null  object        
 2   Product_Category   80000 non-null  object        
 3   Model_Name         80000 non-null  object        
 4   Condition          80000 non-null  object        
 5   Launch_Price_USD   80000 non-null  int64         
 6   Launch_Price_INR   80000 non-null  int64         
 7   Current_Price_USD  80000 non-null  float64       
 8   Current_Price_INR  80000 non-null  float64       
 9   Discount_Pct       80000 non-null  float64       
 10  Sale_Event         6649 non-null   object        
 11  Stock_Status       80000 non-null  object        
 12  Rating             80000 non-null  float64       
 13  Reviews_Count      80000 non-null  int64         
dtypes: dat

In [136]:
df.isnull().Sale_Event.sum()

np.int64(73351)

In [137]:
df["Sale_Event"]=df["Sale_Event"].fillna("No_Sale_Event")

In [138]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80000 entries, 0 to 79999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Date               80000 non-null  datetime64[ns]
 1   Platform           80000 non-null  object        
 2   Product_Category   80000 non-null  object        
 3   Model_Name         80000 non-null  object        
 4   Condition          80000 non-null  object        
 5   Launch_Price_USD   80000 non-null  int64         
 6   Launch_Price_INR   80000 non-null  int64         
 7   Current_Price_USD  80000 non-null  float64       
 8   Current_Price_INR  80000 non-null  float64       
 9   Discount_Pct       80000 non-null  float64       
 10  Sale_Event         80000 non-null  object        
 11  Stock_Status       80000 non-null  object        
 12  Rating             80000 non-null  float64       
 13  Reviews_Count      80000 non-null  int64         
dtypes: dat

In [139]:
df.isnull().Sale_Event.sum()

np.int64(0)

In [140]:
df["Product_Category"].value_counts()

Product_Category
iPhone    28589
Mac       18020
Watch     17865
iPad      15526
Name: count, dtype: int64

In [141]:
df["Stock_Status"].value_counts()

Stock_Status
In Stock        55034
Out of Stock    13475
Low Stock       11491
Name: count, dtype: int64

In [142]:
df.describe()

,Date,Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR,Discount_Pct,Rating,Reviews_Count
count,80000,80000.000000,80000.000000,80000.000000,80000.000000,80000.000000,80000.000000,80000.000000
mean,2024-09-14 22:14:06.359999744,963.959125,94467.994250,782.769855,74628.342439,21.418826,4.450099,2405.995925
min,2020-09-19 00:00:00,329.000000,32242.000000,109.930000,9157.680000,-2.000000,3.800000,10.000000
25%,2023-11-03 00:00:00,599.000000,58702.000000,432.930000,41686.835000,6.700000,4.300000,896.000000
50%,2024-12-22 00:00:00,799.000000,78302.000000,699.740000,67324.020000,21.300000,4.400000,1893.500000
75%,2025-10-27 00:00:00,1199.000000,117502.000000,989.112500,96568.117500,36.800000,4.700000,3383.000000
max,2026-07-31 00:00:00,1999.000000,195902.000000,2038.970000,203668.710000,73.100000,4.900000,11526.000000
std,NaN,470.086413,46068.468464,461.673970,45117.872172,16.697310,0.269782,1973.630852


In [143]:
(df["Discount_Pct"]<0).sum()

np.int64(8669)

In [144]:
df.loc[df["Discount_Pct"]<0,"Discount_Pct"]=0

In [145]:
df.describe()

,Date,Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR,Discount_Pct,Rating,Reviews_Count
count,80000,80000.000000,80000.000000,80000.000000,80000.000000,80000.000000,80000.000000,80000.000000
mean,2024-09-14 22:14:06.359999744,963.959125,94467.994250,782.769855,74628.342439,21.529770,4.450099,2405.995925
min,2020-09-19 00:00:00,329.000000,32242.000000,109.930000,9157.680000,0.000000,3.800000,10.000000
25%,2023-11-03 00:00:00,599.000000,58702.000000,432.930000,41686.835000,6.700000,4.300000,896.000000
50%,2024-12-22 00:00:00,799.000000,78302.000000,699.740000,67324.020000,21.300000,4.400000,1893.500000
75%,2025-10-27 00:00:00,1199.000000,117502.000000,989.112500,96568.117500,36.800000,4.700000,3383.000000
max,2026-07-31 00:00:00,1999.000000,195902.000000,2038.970000,203668.710000,73.100000,4.900000,11526.000000
std,NaN,470.086413,46068.468464,461.673970,45117.872172,16.549541,0.269782,1973.630852


In [146]:
(df["Discount_Pct"]<0).sum()

np.int64(0)

In [147]:
df.to_csv("apple_products_pricing cleaned.csv", index = False)

## Day 5- Work completed
- Reviewed the Day 4 validation findings.
- Created the Data cleaning plan (cleaning plan v1)
- Identified 73,351 missing values in the 'sale_event' column.
- Filled missing values in the 'sale_event' column with 'No_sale_event'.
- Checked the 'Discount_pct' column and identified negative values.
- Updated negative 'Discount_pct' values according to the cleaning plan.
- Verified that 'Launch_Price_USD','Launch_Price_INR','Current_Price_USD','Current_Price_INR' contained valid values and required no changes.
- Veified the cleaned dataset using pandas validation checks.

## Cleaning Decisions

| Column | Issue | Action |
|--------|-------|--------|
| Date | Object data type | Convert to datetime |
| Sale_Event | Missing values | Fill with "No Sale Event" |
| Discount_Pct | Negative values | Replace with 0 |
| Launch_Price_USD | No issue | No action |
| Launch_Price_INR | No issue | No action |
| Current_Price_USD | No issue | No action |
| Current_Price_INR | No issue | No action |
| Product_Category | No issue | No action |
| Stock_Status | No issue | No actio
### Cleaning Plan v1 completed.

In [151]:
df["Date"]=pd.to_datetime(df["Date"],dayfirst=True)

In [152]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80000 entries, 0 to 79999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Date               80000 non-null  datetime64[ns]
 1   Platform           80000 non-null  object        
 2   Product_Category   80000 non-null  object        
 3   Model_Name         80000 non-null  object        
 4   Condition          80000 non-null  object        
 5   Launch_Price_USD   80000 non-null  int64         
 6   Launch_Price_INR   80000 non-null  int64         
 7   Current_Price_USD  80000 non-null  float64       
 8   Current_Price_INR  80000 non-null  float64       
 9   Discount_Pct       80000 non-null  float64       
 10  Sale_Event         80000 non-null  object        
 11  Stock_Status       80000 non-null  object        
 12  Rating             80000 non-null  float64       
 13  Reviews_Count      80000 non-null  int64         
dtypes: dat

In [153]:
df.head()

,Date,Platform,Product_Category,Model_Name,Condition,Launch_Price_USD,Launch_Price_INR,Current_Price_USD,Current_Price_INR,Discount_Pct,Sale_Event,Stock_Status,Rating,Reviews_Count
0,2020-09-19,Flipkart,Watch,Apple Watch Series 6 (44mm),New,429,42042,435.81,43322.41,0.0,No_Sale_Event,In Stock,4.7,40
1,2020-09-20,Flipkart,Watch,Apple Watch Series 6 (44mm),New,429,42042,436.49,42320.43,0.0,No_Sale_Event,Out of Stock,4.6,84
2,2020-09-23,Amazon,Watch,Apple Watch Series 6 (44mm),New,429,42042,422.73,40879.36,1.5,No_Sale_Event,In Stock,4.4,110
3,2020-09-23,Amazon,Watch,Apple Watch Series 6 (44mm),New,429,42042,425.00,42008.70,0.9,No_Sale_Event,In Stock,4.8,111
4,2020-09-24,Amazon,Watch,Apple Watch Series 6 (44mm),New,429,42042,436.22,41984.28,0.0,No_Sale_Event,In Stock,4.7,35
